src/style-spec/expression/definitions/in.js

```diff
@@ -33,18 +41,28 @@ class In implements Expression {
     type: Type;
     needle: Expression;
     haystack: Expression;
+    fromIndex: ?Expression;
 
-    constructor(needle: Expression, haystack: Expression) {
-        this.type = BooleanType;
+    constructor(type: Type, needle: Expression, haystack: Expression, fromIndex?: Expression) {
+        this.type = type;
         this.needle = needle;
         this.haystack = haystack;
+        this.fromIndex = fromIndex;
     }
 
     static parse(args: $ReadOnlyArray<mixed>, context: ParsingContext) {
-        if (args.length !== 3) {
+        const name: string = (args[0]: any);
+        assert(types[name], name);
+
+        if (name === 'in' && args.length !== 3) {
             return context.error(`Expected 2 arguments, but found ${args.length - 1} instead.`);
         }
 
+        if (name === 'index-of' && (args.length <= 2 ||  args.length >= 5))
```

Comment:
I was actually suggesting an inheritance based refactor rather than a runtime check based one.

Reply:
Ahh that sucks because that what I initially did but then swapped it to this method which is how it is in [`./coercion.js`](src/style-spec/expression/definitions/coercion.js).

There is also another method used in [`./comparison.js`](src/style-spec/expression/definitions/comparison.js) which uses a function to create the class.

I didn't want to add a third method of doing it so I went this way, probably should have gotten some clarification before I went and did these changes!

Reply:
Ah, I was unaware of those patterns in those files. They seem like a pattern carried over from ES5, but I'm not 100% certain on this. @ryanhamley do you have any thoughts?

Reply:
No problems at all, happy to modify to whatever the mapbox team thinks fits best. I see at least three options:

1. As is, used in `coercion` - single class and check `arg[0]` to change function return (will admit, it is a bit messy..)
2. Factory method used in `comparision`
3. New pattern for expressions using a base class plus  inheritances classes for `index-of` and `in`

Potentially as a much larger refactor you could change the `Definition` type that goes into `CompoundExpression.register()` to accept generic return types, or overload with a return type and then you wouldn't need custom definitions for `index-of`, `slice` and `in` 

Either way, we can wait for @ryanhamley to provide some input and I can have enough crack at this PR tonight.

Reply:
I think `coercion` is a bit of a special case that's mostly used internally. Instead of combining `in` and `index-of` or having `index-of` inherit from `in`, I think it makes more sense to move common functions such as `isComparableType` to a single file. It may make sense to put them in `expression/types.js` since there are already some shared type checking functions in that file.

Reply:
@ryanhamley & @arindam1993 thanks both for your input. I've refactored with the latest commits. I've moved the functions that check the types into `expression/types.js` and updated the three expressions to use these common functions.
I haven't created a base class that the other extends but if that's what we want to do then I'm happy to do another refactor.